monai/bundle/workflows.py

```diff
@@ -0,0 +1,681 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import os
+import time
+from abc import ABC, abstractmethod
+from logging.config import fileConfig
+from pathlib import Path
+from typing import Any, Sequence
+
+import torch
+
+from monai.apps.utils import get_logger
+from monai.bundle.config_parser import ConfigParser
+from monai.bundle.utils import DEFAULT_EXP_MGMT_SETTINGS
+from monai.engines import Evaluator, Trainer
+from monai.inferers import Inferer
+from monai.transforms import Transform
+
+__all__ = ["BundleWorkflow", "ConfigWorkflow", "ConfigTrainWorkflow", "ConfigInferWorkflow"]
+
+logger = get_logger(module_name=__name__)
+
+
+class BundleWorkflow(ABC):
+    """
+    Base class for the workflow specification in bundle.
+
+    """
+
+    @abstractmethod
+    def initialize(self, *args: Any, **kwargs: Any) -> Any:
+        raise NotImplementedError(f"subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def run(self, *args: Any, **kwargs: Any) -> Any:
+        raise NotImplementedError(f"subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def finalize(self, *args: Any, **kwargs: Any) -> Any:
+        raise NotImplementedError(f"subclass {self.__class__.__name__} must implement this method.")
+
+
+class TrainProperties:
```

Comment:
I'm a little wary of defining such large and rigid class structures like this which make a lot of assumptions about the members of a bundle. This also creates a lot of work for the user to implement. I would suggest we look into a more dynamic definition which uses `__getattr__` and other special methods to implement something like this which is mostly analogous to a [namespace](https://docs.python.org/3/library/types.html#types.SimpleNamespace). 

We can mix and match properties with this idea as well as shown below, what we need then is a list of known symbol names (device, network, trainer, etc.) that we expect clients to look for (whether they're there or not). This list of names would take the place of the list of properties here but with magic methods provide the same functionality. I feel this would be more dynamic because new names could be added to this and there's no requirement that the user implement anything, they could pass a dictionary mapping certain names to value or some such mechanism.

```python
class AttrHolder:
    def __init__(self,values):
        self.values=values
        self._bar=1
        
    @property
    def bar(self):return self._bar

    @bar.setter
    def bar(self,i):
        self._bar=i

    def __getattr__(self, name):
        if name in self.values:
            return self.values[name]
        
        return self.__getattribute__(name)
    
ah=AttrHolder({'foo':123})

print(ah.foo,ah.bar)
ah.bar=3
print(ah.foo,ah.bar)
```

Reply:
I have the same comment, perhaps the properties can also be maintained in a separate dictionary, it's slightly more readable:

```py
train_basic_properties = {
    "bundle_root": None,
    "max_epochs": 1,
    "val_key_metric": 0.0,
    "val_postprocessing": None,
}


class Attr:
    def __init__(self):
        for k, v in train_basic_properties.items():
            setattr(self, k, v)


print(Attr().max_epochs)
```

maybe there are better approaches in this direction.

Reply:
Hi @ericspod @wyli ,

Thanks for your detailed suggestions and proposal!
Sorry maybe I didn't describe it clearly in the draft code for the `TrainProperties`, the main purpose of this interface is to **force the subclasses to implement the required `get / set` APIs to be compatible with other MONAI world packages**. Currently, all these properties are hard requirements from `MONAI Label`, `MONAI Deploy` and `MONAI FL`.
If we can force subclasses to implement these APIs, then developers of `model-zoo` bundles can directly support `MONAI Label`, `MONAI Deploy` and `MONAI FL` even they want to write their own workflow in python without any JSON / YAML config, and we can add CI tests to check the bundle PRs that whether their workflow script inherits from this `TrainProperties` class.
Seems your proposal solutions added the feature to easily get / set properties but no hard-requirements for subclasses to implement anything? Could you please explain more about it to help me understand?

Thanks in advance.

Reply:
I think that'll be too restrictive, not implementing a few of the 20+ properties in an application shouldn't always be a fatal error in most use cases, right?

Reply:
Hi @wyli ,

I defined the properties according to the doc:
https://github.com/Project-MONAI/model-zoo/blob/dev/CONTRIBUTING.md#preferred-files-and-keys
I think all of them are necessary if the bundle is supposed to be in model-zoo? And already covered by model-zoo CI, the original discussion with MONAI Label, MONAI deploy, MONAI FL is here:
https://github.com/Project-MONAI/model-zoo/issues/52
Do you mean we should ask MONAI Label, MONAI deploy, MONAI FL teams to clearly mark the properties with "required" or "optional"?
And for the "required" properties, do you have other ideas if not defined as base class interfaces in this PR?

Thanks.

Reply:
> I think all of them are necessary if the bundle is supposed to be in model-zoo? And already covered by model-zoo CI

This statement is not trur... The current design doesn't strictly require all the properties https://github.com/Project-MONAI/model-zoo/blob/6f418a4571a064eb0fb698cee5c3d81ba820f21a/ci/verify_bundle.py#L137

Reply:
@Nic-Ma I can see your motivation here to ensure that a bundle has the features that we need for these different applications. The problems I see with this approach are that these requirements are not strict but if they were they'd be too restrictive, I can imagine bundles such as those for latent diffusion models that will have extra components not mentioned here. Other bundles may not have certain components you want to be required, whether these can be used or not in these different contexts would depend a lot on what it does but I still see situations that a long list of requirements will cause difficulty. 

The other issue I see is that having a long list of methods a user must implement is quite onerous and more difficult for us to modify over time. What I wanted to demonstrate with my example code is that Python provides us with much more dynamic ways of defining the sort of structures that you want to have here, and we then use some other mechanism to state and enforce required properties. We could have a `Enum` storing names that should be present and a string stating what these are, attributes would be set for each which can be overridden by the user individually:

```python
class RequiredTrainIDs(Enum):
    # (default name, description) pairs
    device = "device", "Device to move all tensors onto"
    train_dataset = "train#dataset", "Dataset storing training data"
# and so forth

class Properties:
    def __init__(self, prop_enum, **overrides):
        self.values={}
        for prop for prop_enum:
            pname=prop.name
            self.values[pname] = overrides.get(pname, pname.value[0]))
        # check every definition we need is present in the loaded bundle and issue warnings for what's missing

    def __getattr__(self, name):
        if name in self.values:
            return self.values[name]  # or resolve the value in the bundle here if it's present, raise error if not
        
        return self.__getattribute__(name)

    # __setattr__ would do the analogous thing here of setting values in the bundle (and raise error if trying to create something new?)
```

This is quite similar to your existing implementation and @wyli's suggestion, where every getter calls `get_content` and every setter calls `set_content`. We can use Python's dynamic way of controlling behaviour to avoid making huge lists of nearly identical methods.

What you want to achieve is fix requirements by defining methods which must be implemented, instead I'm suggesting to fix requirements through the members of an `Enum` which state their default name in the bundle, a default value perhaps, a description that can be included in error outputs, and other properties if needed. The `Properties` class will appear to have these members so the API is going to be very close to yours but with less effort, plus you have the virtue of being able to refer to members of the enumerations by a name rather than an index in a list.

Given all this I do wonder if embedding requirements like this deep in our code is a good idea anyway, perhaps we should have separate schema files for different classes of bundles stating what members they need to have? It's good practice from what I see to not encode data within code so that it's more visible (documened) to users and more easily changed.

Reply:
Hi @ericspod ,

Thanks for your very detailed explanation and example code, I got your idea to reduce the long list functions.
There are 2 points we want to restrict a bundle:
(1) If a bundle defines everything in a JSON or YAML config as most model-zoo bundles, we provide the default `TrainWorkflow` to check the required properties and control the read / write behaviors.
(2) If a bundle defines in python code directly, we want to provide a workflow interface for the developer to inherit and develop his own `CustomizedTrainWorkflow` and we just read / write it in the same way as config-based bundles, because it must implement the required properties. See this example: https://github.com/Nic-Ma/MONAI/blob/5821-bundle-interface/tests/test_bundle_infer_workflow.py#L47.

I think your example mainly solves (1)? For (2), how to restrict or check customized workflow? Do you expect developers to inherit from your class `Properties`? I still don't quite understand your proposal about this part.

Thanks in advance.

Reply:
Hi @Nic-Ma! What I'm thinking is that `BundleWorkflow` and its descendents would mostly stay the same as they are except the data about what is needed in a workflow is moved to separate JSON files which act as schemas for the types of workflows that you'd have. For example, the values in `required_train_ids` would instead go into a JSON file where there would be an entry for each needed ID along with a description of what this would be. `ConfigTrainWorkflow` would be created with a path/URL to this schema file and would then load the configuration it needs from it, and allow a provided argument dictionary to define overrides the user can specify. The object can then check that the bundle it loads has all the required components. With a `Properties` class that uses `__getattr__` to dynamically define which properties are present this avoids writing the large series of methods that you have here and separates data from code. 

We would be defining a new type of JSON config file that would require its own schema as well, but it would look something like this:

```json
{
"name": "TrainProperties",
"properties": [
    {
        "name": "bundle_root",
        "symbol": "bundle_root",
        "description": "Root path of the bundle",
        "required": true
    },
    {
        "name": "train_transforms",
        "symbol": "train#dataset#transform",
        "description": "Transforms to apply to items from the training dataset",
        "required": false
    }
]
}
```

When creating a workflow object we'd specify this file which would load and fill the `Properties` object:

```python
wf = BundleWorkflow("/path/to/train_properties.json", {"bundle_root": "/path/to/actual/bundle"})
wf.initialize()
wf.run()
wf.finalize()
```

I think this does essentially everything that you propose here except without the boilerplate code and moves the description of what bundles must provide elsewhere. We can perhaps merge `BundleWorkflow` and `Properties` into one class because there isn't a specific implementation for training or inference now (we might also want to us a different word from `Workflow` which is used for the Ignite classes we define). A developer can define `CustomizedTrainWorkflow` and inherit from ours then add what they like to that class. 

I have a very rough sketch of what this would look like which mostly takes the code from `ConfigWorkflow` and integrates the idea of loading this schema. It uses most of the meat of your current implementation but uses the schema file and dictionaries containing required and optional components to perform necessary checks and implement what `Properties` does using `__getattr__` instead:

```python
class BundleWorkflow(ABC):
    def __init__(
        self,
        prop_schema: str,
        meta_file: str | Sequence[str] | None = None,
        config_file: str | Sequence[str] | None = None,
        logging_file: str | None = None,
        tracking: str | dict | None = None,
        **override,
    ) -> None:
        
        # load required and optional properties from schema file, each dict is property name to property symbol
        # eg. required_props will contain "train_transform" -> "train#dataset#transform"
        self.required_props, self.optional_props=self._load_properties_from_file(prop_schema)
                
        if logging_file is not None:
            if not os.path.exists(logging_file):
                raise FileNotFoundError(f"can't find the logging config file: {logging_file}.")
            logger.info(f"set logging properties based on config: {logging_file}.")
            fileConfig(logging_file, disable_existing_loggers=False)

        self.parser = ConfigParser()
        self.parser.read_config(f=config_file)
        if meta_file is not None:
            self.parser.read_meta(f=meta_file)
        
        # the rest key-values in the _args are to override config content
        self.parser.update(pairs=override)
        self.init_id = self.required_props["init"]
        self.run_id = self.required_props["run"]
        self.final_id = self.required_props["finalize"]

        # check that required definitions are present in the bundle by checking the members of required_props
        required_not_found=[n for n,sym in self.required_props.items() if sym not in self.parser]        
        if required_not_found:
            raise ValueError(f"Loaded bundle does not contain the following required definitions: {required_not_found}")

        # set tracking configs for experiment management
        if tracking is not None:
            if isinstance(tracking, str) and tracking in DEFAULT_EXP_MGMT_SETTINGS:
                settings_ = DEFAULT_EXP_MGMT_SETTINGS[tracking]
            else:
                settings_ = ConfigParser.load_config_files(tracking)
            self.patch_bundle_tracking(parser=self.parser, settings=settings_)
            
    def __getattr__(self, name):
        
        if name in self.required_props:
            resolved_name=self.required_props[name]
        elif name in self.optional_props:
            resolved_name=self.optional_props[name]
        else:
            return self.__getattribute__(name)  # getting regular attribute

        return self.get_content(resolved_name)
        
    # __setattr__ would do the analogous thing here of setting values in the bundle with set_content or assign to member value if not in either props dict

    def initialize(self) -> Any:
        # reset the "reference_resolver" buffer at initialization stage
        self.parser.parse(reset=True)
        return self._run_expr(id=self.init_id)

    def run(self) -> Any:
        return self._run_expr(id=self.run_id)

    def finalize(self) -> Any:
        return self._run_expr(id=self.final_id)

    def _run_expr(self, id: str, **kwargs) -> Any:
        return self.parser.get_parsed_content(id, **kwargs) if id in self.parser else None

    def get_content(self, id: str, allow_missing: bool = False):
        if not self.parser.ref_resolver.is_resolved():
            raise RuntimeError("please execute 'initialize' before getting any parsed content.")
        if id not in self.parser:
            if allow_missing:
                return None
            else:
                raise KeyError(f"id '{id}' not in the config.")
        return self.parser.get_parsed_content(id=id)

    def set_content(self, id: str, content: Any, allow_missing: bool = False):
        if id not in self.parser:
            if allow_missing:
                return
            else:
                raise KeyError(f"id '{id}' not in the config.")
        self.parser[id] = content
        # must parse the config again after changing the content
        self.parser.ref_resolver.reset()

    @staticmethod
    def patch_bundle_tracking(parser: ConfigParser, settings: dict):
        """
        Patch the loaded bundle config with a new handler logic to enable experiment tracking features.

        Args:
            parser: loaded config content to patch the handler.
            settings: settings for the experiment tracking, should follow the pattern of default settings.

        """
        # unchanged
```

Let me know if that's making more sense or if I've missed part of what your objectives were here but this seems like the way forward to me.

Reply:
Hi @ericspod ,

Thanks very much for your detailed explanation and example code.
I got your point, maybe for the first version we can simply define the schema in the source code directly.
And my previous concern is how to force "python-defined" bundles to follow the properties? For example, this PR wants to develop a new bundle with python code instead of JSON config: https://github.com/Project-MONAI/model-zoo/pull/262
How to require the developer to implement the python workflow and support our properties? Maybe you can modify this demo code as a simple example? https://github.com/Nic-Ma/MONAI/blob/5821-bundle-interface/tests/test_bundle_infer_workflow.py#L47

Thanks in advance.